File: Tree/102.binary-tree-level-order-traversal.cpp
/*
 * @lc app=leetcode id=102 lang=cpp
 * save queue has vector<int>
 * push vector<int> 
 * [102] Binary Tree Level Order Traversal
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        queue<TreeNode* > queue; //save TreeNode
        vector<int> layer;
        vector<vector<int>> result;
        if(root) queue.push(root);

        while(!queue.empty()){
            int current_len = queue.size(); //ERROR 1
            for(int i = 0; i <current_len;i++){
                TreeNode* temp = queue.front();
                layer.push_back(temp->val);            
                queue.pop();
                if(temp->left) queue.push(temp->left);
                if(temp->right) queue.push(temp->right);
            }
            result.push_back(layer);
            layer.clear();
        }
        return result;
    }
};

File: Tree/Recursion/102.binary-tree-level-order-traversal.java
/*
 * @lc app=leetcode id=102 lang=java
 *
 * [102] Binary Tree Level Order Traversal
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> ans = new ArrayList<>();
        if(root == null) return ans;
        DFS(root,ans,0);
        return ans;
    }
    public void DFS(TreeNode root, List<List<Integer>> ans, int level){
        if(root==null) return;
        if(ans.size()==level){
            List<Integer> curr = new ArrayList<>();
            curr.add(root.val);
            ans.add(curr);
        }else{
            ans.get(level).add(root.val);
        }
        DFS(root.left, ans, level+1);
        DFS(root.right, ans, level+1);
    }
}
// @lc code=end

